### PR TITLE
Improve DMs

### PR DESCRIPTION
**Pull request by lelemmen**
In this PR, I've improved the APIs for the density matrices and their Python bindings. I reverted the notion that a density matrix __is__ a matrix/tensor, since that gave some nasty problems in the Python bindings.

### CODE REVIEW

File: gqcp/include/DensityMatrix/MixedSpinResolved2DMComponent.hpp [hunk 1]
```diff
@@ -117,14 +172,14 @@ class MixedSpinResolved2DMComponent:
         // Depending on the given spin-component, we should either transform the first two, or the second two axes.
         switch (sigma) {
         case Spin::alpha: {
-            const auto temp = T_related_tensor.template einsum<1>("UQ,TUVW->TQVW", *this);
+            const auto temp = T_related_tensor.template einsum<1>("UQ,TUVW->TQVW", this->tensor());
```
Comment (dariatols) [inline gqcp/include/DensityMatrix/MixedSpinResolved2DMComponent.hpp:175]: I like how this small adjustment already looks much more intuitive!

File: gqcp/include/DensityMatrix/Simple1DM.hpp [hunk 2]
```diff
@@ -89,7 +155,7 @@ class Simple1DM:
 
         // The transformation formulas for one-electron operators and 1-DMs are similar, but not quite the same. Instead of using T, the transformation formula for the 1-DM uses T_inverse_transpose. See also (https://gqcg-res.github.io/knowdes/spinor-transformations.html).
         const auto T_related = T.matrix().transpose().inverse();
-        return DerivedDM {T_related.adjoint() * (*this) * T_related};
+        return DerivedDM {T_related.adjoint() * this->matrix() * T_related};
```
Comment (dariatols) [inline gqcp/include/DensityMatrix/Simple1DM.hpp:158]: Same.

**Review by dariatols — APPROVED** [review]

File: gqcp/include/Processing/Properties/expectation_values.hpp [hunk 3]
```diff
@@ -43,11 +43,11 @@ double calculateSpinSquared(const SpinResolved1DM<Scalar>& one_DMs, const SpinRe
...
+        s_squared += one_DMs.alpha().matrix()(p, p);                                        // One-electron partition of S+S_
+        s_squared += (one_DMs.alpha().matrix()(p, p) + one_DMs.beta().matrix()(p, p)) / 4;  // One-electron partition of S^2
         for (size_t q = 0; q < K; q++) {
-            s_squared += -two_DMs.alphaBeta()(p, q, q, p);                                                                                                             // Two-electron partition  S+S_
-            s_squared += (two_DMs.alphaAlpha()(p, p, q, q) + two_DMs.betaBeta()(p, p, q, q) - two_DMs.alphaBeta()(p, p, q, q) - two_DMs.betaAlpha()(p, p, q, q)) / 4;  // Two-electron partition of S^2
+            s_squared += -two_DMs.alphaBeta().tensor()(p, q, q, p);                                                                                                                                        // Two-electron partition  S+S_
+            s_squared += (two_DMs.alphaAlpha().tensor()(p, p, q, q) + two_DMs.betaBeta().tensor()(p, p, q, q) - two_DMs.alphaBeta().tensor()(p, p, q, q) - two_DMs.betaAlpha().tensor()(p, p, q, q)) / 4;  // Two-electron partition of S^2
```
Comment (xdvriend) [inline gqcp/include/Processing/Properties/expectation_values.hpp:50]:
```suggestion
            s_squared += -two_DMs.alphaBeta().tensor()(p, q, q, p);                                                                                                                                        // Two-electron partition  S+S_ .
            s_squared += (two_DMs.alphaAlpha().tensor()(p, p, q, q) + two_DMs.betaBeta().tensor()(p, p, q, q) - two_DMs.alphaBeta().tensor()(p, p, q, q) - two_DMs.betaAlpha().tensor()(p, p, q, q)) / 4;  // Two-electron partition of S^2 .
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 4]
```diff
@@ -747,8 +747,8 @@ class LinearExpansion {
         // Initialize as zero matrices
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:747]:
```suggestion
        // Initialize as zero matrices.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 5]
```diff
@@ -852,7 +852,7 @@ class LinearExpansion {
             }
 
         }  // I_beta loop
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:854]:
```suggestion
        }  // I_beta loop.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 6]
```diff
@@ -875,9 +875,9 @@ class LinearExpansion {
         // Initialize as zero matrices
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:875]:
```suggestion
        // Initialize as zero matrices.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 7]
```diff
@@ -875,9 +875,9 @@ class LinearExpansion {
...
         size_t K = this->onv_basis.alpha().numberOfOrbitals();
 
-        PureSpinResolved2DMComponent<double> d_aaaa = PureSpinResolved2DMComponent<double>::Zero(K);
-        MixedSpinResolved2DMComponent<double> d_aabb = MixedSpinResolved2DMComponent<double>::Zero(K);
-        PureSpinResolved2DMComponent<double> d_bbbb = PureSpinResolved2DMComponent<double>::Zero(K);
+        SquareRankFourTensor<double> d_aaaa = SquareRankFourTensor<double>::Zero(K);
+        SquareRankFourTensor<double> d_aabb = SquareRankFourTensor<double>::Zero(K);
+        SquareRankFourTensor<double> d_bbbb = SquareRankFourTensor<double>::Zero(K);
 
         // ALPHA-ALPHA-ALPHA-ALPHA
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:882]:
```suggestion
        // ALPHA-ALPHA-ALPHA-ALPHA.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 8]
```diff
@@ -1004,8 +1004,8 @@ class LinearExpansion {
 
         // BETA-BETA-ALPHA-ALPHA
         // We know that d^aabb_pqrs = d^bbaa_rspq
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1006]:
```suggestion
        // BETA-BETA-ALPHA-ALPHA.
        // We know that d^aabb_pqrs = d^bbaa_rspq.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 9]
```diff
@@ -1004,8 +1004,8 @@ class LinearExpansion {
...
-        Eigen::array<int, 4> shuffle {2, 3, 0, 1};  // array specifying the axes that should be swapped
-        MixedSpinResolved2DMComponent<double> d_bbaa {d_aabb.Eigen().shuffle(shuffle)};
+        Eigen::array<int, 4> axes {2, 3, 0, 1};  // array specifying the axes that should be swapped
+        MixedSpinResolved2DMComponent<double> d_bbaa {GQCP::SquareRankFourTensor<double>(d_aabb.Eigen().shuffle(axes))};
 
 
         // BETA-BETA-BETA-BETA
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1011]:
```suggestion
        // BETA-BETA-BETA-BETA.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 10]
```diff
@@ -1065,7 +1065,7 @@ class LinearExpansion {
 
         }  // loop over I_beta
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1066]:
```suggestion
        }  // loop over I_beta.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 11]
```diff
@@ -1204,7 +1204,7 @@ class LinearExpansion {
         }
 
         // For seniority-zero linear expansions, we have additional symmetries (two_rdm_aaaa = two_rdm_bbbb, two_rdm_aabb = two_rdm_bbaa)
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1206]:
```suggestion
        // For seniority-zero linear expansions, we have additional symmetries (two_rdm_aaaa = two_rdm_bbbb, two_rdm_aabb = two_rdm_bbaa).
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 12]
```diff
@@ -1223,8 +1223,8 @@ class LinearExpansion {
         size_t K = this->onv_basis.numberOfOrbitals();
         size_t dim = onv_basis.dimension();
 
-        SpinResolved1DMComponent<double> D_aa = SpinResolved1DMComponent<double>::Zero(K);
-        SpinResolved1DMComponent<double> D_bb = SpinResolved1DMComponent<double>::Zero(K);
+        SquareMatrix<double> D_aa = SquareMatrix<double>::Zero(K);
+        SquareMatrix<double> D_bb = SquareMatrix<double>::Zero(K);
 
 
         for (size_t I = 0; I < dim; I++) {  // loop over all addresses (1)
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1230]:
```suggestion
        for (size_t I = 0; I < dim; I++) {  // Loop over all addresses (1).
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 13]
```diff
@@ -1288,7 +1288,7 @@ class LinearExpansion {
             }  // loop over addresses J > I
         }      // loop over addresses I
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1289]:
```suggestion
            }  // Loop over addresses J > I.
        }      // Loop over addresses I.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 14]
```diff
@@ -1303,10 +1303,10 @@ class LinearExpansion {
...
-        PureSpinResolved2DMComponent<double> d_aaaa = PureSpinResolved2DMComponent<double>::Zero(K);
-        MixedSpinResolved2DMComponent<double> d_aabb = MixedSpinResolved2DMComponent<double>::Zero(K);
-        MixedSpinResolved2DMComponent<double> d_bbaa = MixedSpinResolved2DMComponent<double>::Zero(K);
-        PureSpinResolved2DMComponent<double> d_bbbb = PureSpinResolved2DMComponent<double>::Zero(K);
+        SquareRankFourTensor<double> d_aaaa = SquareRankFourTensor<double>::Zero(K);
+        SquareRankFourTensor<double> d_aabb = SquareRankFourTensor<double>::Zero(K);
+        SquareRankFourTensor<double> d_bbaa = SquareRankFourTensor<double>::Zero(K);
+        SquareRankFourTensor<double> d_bbbb = SquareRankFourTensor<double>::Zero(K);
 
         for (size_t I = 0; I < dim; I++) {  // loop over all addresses I
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1311]:
```suggestion
        for (size_t I = 0; I < dim; I++) {  // Loop over all addresses I.
```

File: gqcp/include/QCModel/CI/LinearExpansion.hpp [hunk 15]
```diff
@@ -1520,7 +1520,7 @@ class LinearExpansion {
 
         }  // loop over all addresses I
```
Comment (xdvriend) [inline gqcp/include/QCModel/CI/LinearExpansion.hpp:1521]:
```suggestion
        }  // Loop over all addresses I.
```

File: gqcp/tests/Operator/SecondQuantized/SimpleSQTwoElectronOperator_test.cpp [hunk 16]
```diff
@@ -187,11 +186,12 @@ BOOST_AUTO_TEST_CASE(calculateExpectationValue_behaviour) {
             }
         }
     }
+    const GQCP::Orbital2DM<double> dm {d};
 
     // Initialize a reference value
```
Comment (xdvriend) [inline gqcp/tests/Operator/SecondQuantized/SimpleSQTwoElectronOperator_test.cpp:191]:
```suggestion
    // Initialize a reference value.
```

File: gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_test.cpp [hunk 17]
```diff
@@ -81,7 +81,7 @@ BOOST_AUTO_TEST_CASE(H3_test_1) {
     const double reference_s_y = 0.0;  // an UHF solution
     const double reference_s_z = 0.5;  // an UHF solution
```
Comment (xdvriend) [inline gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_test.cpp:82]:
```suggestion
    const double reference_s_y = 0.0;  // An UHF solution.
    const double reference_s_z = 0.5;  // An UHF solution.
```

File: gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_test.cpp [hunk 18]
```diff
@@ -153,7 +153,7 @@ BOOST_AUTO_TEST_CASE(H3_test_2) {
     const double reference_s_y = 0.0;                     // a true GHF solution
     const double reference_s_z = -4.903573113845816e-05;  // a true GHF solution
```
Comment (xdvriend) [inline gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_test.cpp:154]:
```suggestion
    const double reference_s_y = 0.0;                     // A true GHF solution.
    const double reference_s_z = -4.903573113845816e-05;  // A true GHF solution.
```

File: gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_test.cpp [hunk 19]
```diff
@@ -223,7 +223,7 @@ BOOST_AUTO_TEST_CASE(H3_test_DIIS) {
     const double reference_s_y = 0.0;                 // an UHF solution
     const double reference_s_z = 0.4999999999999999;  // an UHF solution
```
Comment (xdvriend) [inline gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_test.cpp:224]:
```suggestion
    const double reference_s_y = 0.0;                 // An UHF solution.
    const double reference_s_z = 0.4999999999999999;  // An UHF solution.
```

**Review by xdvriend — APPROVED** [review]
I like the changes you made to the API's! This PR is good to go. The suggestions I've left are all just punctuation changes in some of the comment lines.

**Pull request merged**